hetmech/degree_weight.py

```diff
@@ -152,6 +153,116 @@ def dwwc(graph, metapath, damping=0.5, dense_threshold=0, dtype=numpy.float64):
+    arrays = []
+    row_names = None
+    array_dims = []
+    for edge in metapath:
+        rows, cols, adj_mat = metaedge_to_adjacency_matrix(
+            graph, edge, dense_threshold=dense_threshold, dtype=dtype)
+        adj_mat = _degree_weight(adj_mat, damping, dtype=dtype)
+        if row_names is None:
+            row_names = rows
+        arrays.append(adj_mat)
+        array_dims.append(adj_mat.shape[0])
+    array_dims.append(adj_mat.shape[1])
+
+    # Find optimal matrix chain ordering. See https://git.io/vh38o
+    n = len(metapath)
+    m = numpy.zeros((n, n), dtype=numpy.double)
+    ordering = numpy.empty((n, n), dtype=numpy.intp)
+    for l in range(1, n):
+        for i in range(n - l):
+            j = i + l
+            m[i, j] = numpy.inf
+            for k in range(i, j):
+                q = m[i, k] + m[k + 1, j] + array_dims[i] * array_dims[k + 1] * array_dims[j + 1]
+                if q < m[i, j]:
+                    m[i, j] = q
+                    ordering[i, j] = k
+
+    dwwc_matrix = numpy.linalg.linalg._multi_dot(arrays, ordering, 0, n - 1)
+    dwwc_matrix = sparsify_or_densify(dwwc_matrix, dense_threshold)
+    return row_names, cols, dwwc_matrix
+
+
+def dwwc_chain_nomem(graph, metapath, damping=0.5, dense_threshold=0, dtype=numpy.float64):
+    """
+    Equivalent to dwwc_chain, but this does not keep all matrices in memory during
+    computation. https://git.io/vh38o
+    """
+    metapath = graph.metagraph.get_metapath(metapath)
+    row_names = None
+    array_dims = []
+    for edge in metapath:
+        source = edge.source
+        target = edge.target
+        rows = graph.get_node_identifiers(source)
+        array_dims.append(len(rows))
+        if row_names is None:
+            row_names = rows
+    column_names = graph.get_node_identifiers(target)
+    array_dims.append(len(column_names))
+
+    # Find optimal matrix chain ordering. See https://git.io/vh38o
+    n = len(metapath)
+    m = numpy.zeros((n, n), dtype=numpy.double)
+    ordering = numpy.empty((n, n), dtype=numpy.intp)
+    for l in range(1, n):
+        for i in range(n - l):
+            j = i + l
+            m[i, j] = numpy.inf
+            for k in range(i, j):
+                q = m[i, k] + m[k + 1, j] + array_dims[i] * array_dims[k + 1] * array_dims[j + 1]
+                if q < m[i, j]:
+                    m[i, j] = q
+                    ordering[i, j] = k
+
+    def _multi_dot(metapath, order, i, j):
+        """
+        Actually do the multiplication with the given order. Modified from
+        original at https://git.io/vh31f
+        """
+        if i == j:
+            _, _, adj_mat = metaedge_to_adjacency_matrix(
+                graph, metapath[i], dense_threshold=dense_threshold, dtype=dtype)
+            adj_mat = _degree_weight(adj_mat, damping)
+            return adj_mat
+        else:
+            return numpy.dot(_multi_dot(metapath, order, i, order[i, j]),
```

Comment:
Does this work for sparse matrices?

Reply:
It appears to work, but I am switching it to use `@`